### PR TITLE
Created the ability to add projects to vulnerabilities on FossGraphView

### LINKED ISSUE #1970: FossGraph. Create the ability to add projects to vulnerabilities on FossGraphView

### PR DESCRIPTION
**Pull request by Cheshiriks**
* created the ability to add projects to vulnerabilities on FossGraphView

С2332

Closes #1970

### CODE REVIEW

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt [hunk 1]
```diff
@@ -72,4 +73,19 @@ class VulnerabilityController(
     }.map {
         ResponseEntity.ok("Vulnerability was successfully saved")
     }
+
+    @PostMapping("/save-all-projects")
```
Comment (sanyavertolet) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt:77]: It seems to be a little bit complicated endpoint - too many characters

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt [hunk 2]
```diff
@@ -72,4 +73,19 @@ class VulnerabilityController(
...
+        summary = "Update vulnerability.",
+        description = "Update vulnerability.",
+    )
+    @ApiResponse(responseCode = "200", description = "Successfully update vulnerability")
+    fun saveAllProjects(
+        @RequestBody vulnerabilityDtos: List<VulnerabilityProjectDto>,
+    ): Mono<StringResponse> = blockingToMono {
+        vulnerabilityService.saveAllProjects(vulnerabilityDtos)
+    }.map {
+        ResponseEntity.ok("Vulnerability was successfully update")
```
Comment (sanyavertolet) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt:89]:
```suggestion
        ResponseEntity.ok("Vulnerability was successfully updated")
```

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt [hunk 3]
```diff
@@ -72,4 +73,19 @@ class VulnerabilityController(
...
         ResponseEntity.ok("Vulnerability was successfully saved")
     }
+
+    @PostMapping("/save-all-projects")
+    @Operation(
+        method = "POST",
+        summary = "Update vulnerability.",
+        description = "Update vulnerability.",
+    )
+    @ApiResponse(responseCode = "200", description = "Successfully update vulnerability")
```
Comment (sanyavertolet) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt:83]:
```suggestion
    @ApiResponse(responseCode = "200", description = "Successfully updated vulnerability")
```

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt [hunk 4]
```diff
@@ -52,10 +55,37 @@ val fossGraphView: VFC = VFC {
...
     val (vulnerability, setVulnerability) = useState(VulnerabilityDto.empty)
+    val (isUpdateVulnerability, setIsUpdateVulnerability) = useState(false)
+
+    val fetchProject: (VulnerabilityProjectDto) -> Unit = { project ->
+        setVulnerability {
+            it.copy(projects = it.projects.plus(project))
+        }
+        setVulnerabilityProjects {
+            it.plus(project)
+        }
```
Comment (sanyavertolet) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt:73]: Do you want to duplicate `VulnerabilityProject`s info?

**Review by sanyavertolet — APPROVED** [review]

**Review by nulls — APPROVED** [review]

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/VulnerabilityProjectWindow.kt [hunk 5]
```diff
@@ -46,7 +51,7 @@ external interface VulnerabilityProjectWindowProps : Props {
 private fun createVulnerabilityProjectWindow() = FC<VulnerabilityProjectWindowProps> { props ->
 
     val (conflictErrorMessage, setConflictErrorMessage) = useState<String?>(null)
-    val (vulnerabilityProject, setVulnerabilityProject) = useState(VulnerabilityProjectDto.empty)
+    val (vulnerabilityProject, setVulnerabilityProject) = useState(VulnerabilityProjectDto("", "", false, props.vulnerabilityName))
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/VulnerabilityProjectWindow.kt:54]: I think it looks more cumbersome and requires the creation of an additional unnecessary object.

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt [hunk 6]
```diff
@@ -52,10 +55,37 @@ val fossGraphView: VFC = VFC {
...
+        setIsUpdateVulnerability(true)
+        projectWindowOpenness.closeWindow()
+    }
+
+    val enrollRequest = useDeferredRequest {
+        val response = post(
+            url = "$apiUrl/vulnerabilities/save-all-projects",
+            headers = jsonHeaders,
+            body = Json.encodeToString(vulnerabilityProjects.toList()),
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt:73]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt (hunk 2 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt (hunk 1 above)
Comment (Cheshiriks) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt]: Fixed

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt:73]: _Vulnerability_ is needed to display on the frontend, and _VulnerabilityProjects_ is passed to the backend.

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt [hunk 7]
```diff
@@ -74,18 +74,18 @@ class VulnerabilityController(
...
 
-    @PostMapping("/save-all-projects")
+    @PostMapping("/save-projects")
     @Operation(
         method = "POST",
         summary = "Update vulnerability.",
         description = "Update vulnerability.",
     )
-    @ApiResponse(responseCode = "200", description = "Successfully update vulnerability")
+    @ApiResponse(responseCode = "200", description = "Successfully updateв vulnerability")
```
Comment (kgevorkyan) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt:83]:
```suggestion
    @ApiResponse(responseCode = "200", description = "Successfully updated vulnerability")
```

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt [hunk 8]
```diff
@@ -72,4 +73,19 @@ class VulnerabilityController(
...
         ResponseEntity.ok("Vulnerability was successfully saved")
     }
+
+    @PostMapping("/save-projects")
+    @Operation(
+        method = "POST",
+        summary = "Update vulnerability.",
+        description = "Update vulnerability.",
+    )
+    @ApiResponse(responseCode = "200", description = "Successfully updateв vulnerability")
```
Comment (sanyavertolet) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt:83]:
```suggestion
    @ApiResponse(responseCode = "200", description = "Successfully updated vulnerability")
```

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt (hunk 7 above)
Comment (nulls) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/vulnerability/VulnerabilityController.kt:83]: https://yandex.ru/soft/punto/win/ :)

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt]: I don't see changes

[the remark this answers is not in the document]
Comment (nulls) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/fossgraph/FossGraphView.kt]: yeap, my bad, looked at old commit

**Pull request merged**